subsys/power/device_pm.c

```diff
@@ -0,0 +1,204 @@
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	target_state = dev->pm->target_state;
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		dev->pm->fsm_state = DEVICE_PM_FSM_STATE_RESUMING;
+	} else {
+		dev->pm->fsm_state = DEVICE_PM_FSM_STATE_SUSPENDING;
+	}
+	fsm_state = dev->pm->fsm_state;
+	k_sem_give(&dev->pm->lock);
+
+	ret = device_set_power_state(dev, target_state,
+				     device_pm_callback, &fsm_state);
+	if (ret < 0) {
+		LOG_ERR("Set power state(%d) error(%d)\n", target_state, ret);
+	}
+}
+
+static int device_pm_request(struct device *dev,
+			     u32_t target_state, u32_t pm_flags)
+{
+	int ret = 0;
+
+	__ASSERT((target_state == DEVICE_PM_ACTIVE_STATE) ||
+			(target_state == DEVICE_PM_SUSPEND_STATE),
+			"Invalid device PM state requested");
+
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		if (atomic_inc(&dev->pm->usage) < 0) {
+			goto req_out;
+		}
+	} else {
+		if (atomic_dec(&dev->pm->usage) > 1) {
+			goto req_out;
+		}
+	}
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (!dev->pm->enable) {
+		k_sem_give(&dev->pm->lock);
+		goto req_out;
+	}
+
+	if (dev->pm->target_state == target_state) {
+		if ((pm_flags & DEVICE_PM_ASYNC) &&
+			(pm_to_fsm_state(target_state) ==
+				dev->pm->fsm_state)) {
+			/* Notify the driver through signal */
+			k_poll_signal_raise(&dev->pm->signal,
+					dev->pm->target_state);
+		}
+		k_sem_give(&dev->pm->lock);
+		goto req_out;
+	}
+
+	dev->pm->target_state = target_state;
+	k_work_submit(&dev->pm->work);
+
+	k_sem_give(&dev->pm->lock);
+
+	/* Incase of Sync request wait for completion event */
+	if (!(pm_flags & DEVICE_PM_ASYNC)) {
+		u32_t signaled;
+		int result;
+
+		ret = k_poll(&dev->pm->event, 1, K_FOREVER);
+		if (ret < 0 || dev->pm->event.state != K_POLL_STATE_SIGNALED) {
+			ret = -EIO;
+		} else {
+			k_poll_signal_check(&dev->pm->signal,
+						&signaled, &result);
+			if (result != target_state) {
+				ret = -EIO;
+			}
+		}
+	}
+
+req_out:
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_ACTIVE_STATE, DEVICE_PM_ASYNC);
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_ACTIVE_STATE, 0);
+}
+
+int device_pm_put(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_SUSPEND_STATE, DEVICE_PM_ASYNC);
+}
+
+int device_pm_put_sync(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_SUSPEND_STATE, 0);
+}
+
+void device_pm_enable(struct device *dev)
+{
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->enable = true;
+
+	/* During the driver init, device can set the
+	 * PM state accordingly. For later cases we need
+	 * to check the usage and set the device PM state.
+	 */
+	if (!dev->pm->dev) {
+		dev->pm->dev = dev;
+		dev->pm->fsm_state = DEVICE_PM_FSM_ACTIVE;
+		dev->pm->target_state = DEVICE_PM_ACTIVE_STATE;
+		k_work_init(&dev->pm->work, pm_work_handler);
+	} else {
+		if (dev->pm->usage) {
+			ret = device_set_power_state(dev,
```

Comment:
This should go through FSM too. Managing state from single place will make the code simpler and easier to maintain.

Reply:
Ok